### PR TITLE
[resolve #18] Add C++ allocator (partial) 

### LINKED ISSUE #18: Add C++ allocator
Currently, we expose a very low-level interface for allocating memory - as shown in this documentation:

```cpp
  rdmalib::Buffer<char> in(opts.input_size, rdmalib::functions::Submission::DATA_HEADER_SIZE), out(opts.input_size);
  in.register_memory(executor._state.pd(), IBV_ACCESS_LOCAL_WRITE);
  out.register_memory(executor._state.pd(), IBV_ACCESS_LOCAL_WRITE | IBV_ACCESS_REMOTE_WRITE);
```

While `rdmalib::Buffer` is applicable, we need a higher-level concept based on the CX++ `std::allocator` to integrate RDMA-managed memory into user applications.

- [ ] Add allocator implementation in `rfaaslib`.
- [ ] Encapsulate the memory registration in `rdmalib` - provide generic enums for local/remote read/write ops + atomics.
- [ ] Support in allocator registration and deregistration of memory.
- [ ] Add test demonstrating standard memory allocation.
- [ ] Add test demonstrating allocation with `std::vector` and our custom allocator.

### PR DESCRIPTION
**Pull request by William-Mou**
## Story

This PR is open for issue #18, including the following features.

  - [x]  Add allocator implementation in rfaaslib.
  - [x] Encapsulate the memory registration in rdmalib 
    - [ ] provide generic enums for local/remote read/write ops + atomics.
  - [x] Support in allocator registration 
    - [ ] and deregistration of memory. (not test yet)
  - [x] Add test demonstrating standard memory allocation.
  - [ ] Add test demonstrating allocation with std::vector and our custom allocator.

## Achievements

- Defines `rdma_allocator.hpp`, which provides the allocate and deallocate functions.
- Use `rfaas::RdmaAllocator<rdmalib::Buffer<char> > rdmaAllocator(executor);` to instantiate the allocator object.
- `rdmaAllocator.allocate(size_t size, int access)` intuitively allocates memory space for a buffer.

## Test
- By embedding `std::cout` in the allocator, you can test the program (eg: `warm_benchmarker` ) to run as expected.
```
warm_benchmarker --config benchmark.json --device-database client_devices.json --name empty --functions /home/mou/rFaaS/build-test/examples/libfunctions.so --executors-database executors_database.json -s 0xFFF
```

## TODO
- Complete the tasks not checked above

### CODE REVIEW

File: rfaas/include/rfaas/rdma_allocator.hpp [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+#ifndef __RFAAS_RDMA_ALLOCATOR_HPP__
+#define __RFAAS_RDMA_ALLOCATOR_HPP__
+
+#include <cstddef>
+#include <rdmalib/buffer.hpp>
+#include <rfaas/executor.hpp>
+
+namespace rfaas {
+  template<typename T>
```
Comment (mcopik) [inline rfaas/include/rfaas/rdma_allocator.hpp:13]: I think the allocator should be of type `T`, and then always return `rdmalib::Buffer` object. If you look at the [C++ allocator concept](https://en.cppreference.com/w/cpp/named_req/Allocator), then I think we should use `A::pointer` as `rdmalib::Buffer`.

File: rfaas/include/rfaas/rdma_allocator.hpp [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+  public:
+    inline explicit RdmaAllocator(const executor &executor) noexcept: _executor(executor) {}
+
+    // inline T *allocate(const std::size_t &, const int &, int = 0);
+    inline T *allocate(const std::size_t &size, const int &access, int header=0) {
+      if (size > std::size_t(-1) / sizeof(T))
+        throw std::bad_alloc();
+
+      auto buffer = new rdmalib::Buffer<char>(size, header);
```
Comment (mcopik) [inline rfaas/include/rfaas/rdma_allocator.hpp:26]: I don't think we need dynamic allocation for the buffer object itself :-) Once we properly define the pointer type.

**Review by mcopik — CHANGES_REQUESTED** [review]

**Comment by William-Mou** [comment]
Hi @mcopik, 
I have updated my code to have a structure similar to the documentation, which may help with communication 😊

**Comment by mcopik** [comment]
@William-Mou Thanks! I will try to test this and merge in the coming days :)

**Comment by William-Mou** [comment]
I proposed a better solution in #30, so #28 has been closed.

**Pull request closed without merging**